h/static/scripts/build-thread.js

```diff
@@ -0,0 +1,295 @@
+'use strict';
+
+/** Default state for new threads, before applying filters etc. */
+var DEFAULT_THREAD_STATE = {
+  /** The Annotation which is displayed by this thread. */
+  annotation: undefined,
+  /** The parent Thread */
+  parent: undefined,
+  /** True if replies to this annotation are hidden. */
+  collapsed: false,
+  /** True if this annotation matches the current filters. */
+  visible: true,
+  /** Replies to this annotation. */
+  children: [],
+  /**
+    * The total number of children of this annotation,
+    * including any which have been hidden by filters.
+    */
+  totalChildren: 0,
+};
+
+/**
+ * Returns a persistent identifier for an Annotation.
+ * If the Annotation has been created on the server, it will have
+ * an ID assigned, otherwise we fall back to the local-only '$$tag'
+ * property.
+ */
+function id(annotation) {
+  return annotation.id || annotation.$$tag;
+}
+
+/**
+ * Creates a thread of annotations from a list of annotations.
```

Comment:
TBH I think this is so much simpler than jwz that I don't think it makes sense to describe it as being the JWZ algorithm any more. Re-reading the JWZ docs though I realized that one difference introduced here is in what happens if an annotation with replies is deleted whilst the user has the sidebar open. In this branch, the reply becomes a top-level annotation and ends up behaving like a page note. In master, there is a placeholder "Message was deleted" annotation. In both cases the replies can be edited and new replies can be created, which are then immediately 'lost' when the page is reloaded. I filed https://trello.com/c/9boX9ToY/